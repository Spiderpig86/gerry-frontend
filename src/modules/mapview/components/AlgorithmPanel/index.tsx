import * as React from 'react';

import { Row, Button, ButtonGroup, Form} from 'react-bootstrap';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { fas, faPlay, faPause, faSquare } from '@fortawesome/free-solid-svg-icons';


export class AlgorithmPanel extends React.PureComponent {
    render() {
        return (
            <Row className={'d-flex w-100'} style={{ position: 'absolute', bottom: 0, padding: '1rem', justifyContent: 'space-between' }}>
                <ButtonGroup>
                    <Button><FontAwesomeIcon icon={faPlay} /></Button>
                    <Button><FontAwesomeIcon icon={faPause} /></Button>
                    <Button><FontAwesomeIcon icon={faSquare} /></Button>
                </ButtonGroup>
                <Form.Group className="row form-group d-flex align-items-center justify-content-flex-end">
                    <Form.Check custom type={'checkbox'} id={'intermediateResultsCheckbox'} label={'Display Intermediate Results'} />
                </Form.Group>
            </Row>
        );
    }
}
