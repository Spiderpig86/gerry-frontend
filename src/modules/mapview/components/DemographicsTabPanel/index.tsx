import * as React from 'react';
import { IDemographics } from '../../../../models';

export interface IDemographicsTabProps {
    hispanicDemographics: IDemographics;
    nonHispanicDemographics: IDemographics;
}

export class DemographicsTabPanel extends React.PureComponent<{}, {}> {
    render() {
        return (
            <div style={{ maxHeight: '80vh', overflow: 'auto', padding: '0 1.5rem' }}>
                <br />
                <h4>Demographics</h4>
                <p>Non-Hispanic White:</p>
                <p>Non-Hispanic African Americans:</p>
                <p>Non-Hispanic Native Americans:</p>
                <p>Non-Hispanic Asian:</p>
                <p>Non-Hispanic Pacific Islander:</p>
                <p>Non-Hispanic Other:</p>
                <p>Non-Hispanic Biracial:</p>
                <br />
                <p>Hispanic:</p>
                <p>Hispanic White:</p>
                <p>Hispanic African Americans:</p>
                <p>Hispanic Native Americans:</p>
                <p>Hispanic Asian:</p>
                <p>Hispanic Pacific Islander:</p>
                <p>Hispanic Other:</p>
                <p>Hispanic Biracial:</p>
            </div>
        )
    }
}