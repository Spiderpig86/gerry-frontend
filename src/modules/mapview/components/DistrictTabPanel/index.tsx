import * as React from 'react';

import { Dropdown, DropdownButton, Image, Table } from 'react-bootstrap';
import { StatisticsAccordionComponent } from '../StatisticsAccordionComponent';
import { Card } from 'react-bootstrap';
import { StateEnum, ICluster, PartyEnum, ClusterCount, MapFilterEnum } from '../../../../models';
import { Placeholder } from '../../../../global_components';
import { Coloring } from '../../../../libs/coloring';
import { EnumNameMapper } from '../../../../libs/enum-name';

interface DistrictTabPanelProps {
    selectedState: StateEnum;
    selectedOldDistrictId: string;
    selectedNewDistrictId: string;
    oldClusters: Map<string, ICluster>;
    newClusters: Map<string, ICluster>;
    mapFilter: MapFilterEnum;
    coloring: Coloring;
    oldClusterCount: ClusterCount;
    newClusterCount?: ClusterCount;
}

interface DistrictTabPanelState {
    oldSortedKeys: string[];
    newSortedKeys: string[];
    selectedOldDistrictId: string;
    oldDistrictData: ICluster;
    selectedNewDistrictId: string;
    newDistrictData: ICluster;
}

export class DistrictTabPanel extends React.PureComponent<DistrictTabPanelProps, DistrictTabPanelState> {
    state = {
        oldSortedKeys: [],
        newSortedKeys: [],
        selectedOldDistrictId: '',
        oldDistrictData: null,
        selectedNewDistrictId: '',
        newDistrictData: null
    };

    componentWillMount() {
        if (this.props.oldClusters) {
            this.setState({
                oldSortedKeys: Array.from(this.props.oldClusters.keys()).sort((a, b) => Number(a) - Number(b))
            });
        }

        if (this.props.newClusters) {
            this.setState({
                newSortedKeys: Array.from(this.props.newClusters.keys()).sort((a, b) => Number(a) - Number(b))
            });
        }

        if (this.props.selectedOldDistrictId !== '0') {
            this.selectOldDistrictData(this.props.selectedOldDistrictId.toString());
        }
        if (this.props.selectedNewDistrictId !== '0') {
            this.selectNewDistrictData(this.props.selectedNewDistrictId.toString());
        }
    }

    async componentWillReceiveProps(newProps: DistrictTabPanelProps) {
        if (newProps.oldClusters.size === 0 || newProps.newClusters.size === 0) {
            this.resetDistrictResults();
        }
        this.setState({
            oldSortedKeys: Array.from(newProps.oldClusters.keys()).sort((a, b) => Number(a) - Number(b))
        });
        this.setState({
            newSortedKeys: Array.from(newProps.newClusters.keys()).sort((a, b) => Number(a) - Number(b))
        });

        if (this.props.selectedOldDistrictId !== '0') {
            this.selectOldDistrictData(this.props.selectedOldDistrictId.toString());
        }
        if (this.props.selectedNewDistrictId !== '0') {
            this.selectNewDistrictData(this.props.selectedNewDistrictId.toString());
        }
        
    }

    render() {
        if (this.props.selectedState === StateEnum.NOT_SET) {
            return (
                <Placeholder
                    loading={false}
                    title="No state selected."
                    subtitle="Select a state to view data."
                ></Placeholder>
            );
        }

        if (!this.props.oldClusters || this.state.oldSortedKeys.length === 0) {
            return <Placeholder loading={true} title="" subtitle=""></Placeholder>;
        }
        return (
            <div style={{ padding: '0 1.5rem' }}>
                <br />
                <h4>District Statistics</h4>
                <h6>Old District Statistics</h6>

                {this.state.oldDistrictData && (
                    <Card
                        className="my-3"
                        style={{
                            borderLeft:
                                this.state.oldDistrictData.incumbent.party === PartyEnum.REPUBLICAN
                                    ? '10px solid #7e3131'
                                    : '10px solid #2f5fa1'
                        }}
                    >
                        <Card.Body>
                            <div className="message">
                                <h5>Incumbent: {this.state.oldDistrictData.incumbent.name}</h5>
                                <h6>Party: {EnumNameMapper.getPartyName(this.state.oldDistrictData.incumbent.party)}</h6>
                            </div>
                        </Card.Body>
                    </Card>
                )}
                <br />
                <div className="d-flex align-items-center">
                    <div
                        className="mr-2"
                        style={{
                            width: '28px',
                            height: '28px',
                            backgroundColor: this.state.selectedOldDistrictId
                                ? this.props.coloring.colors[parseInt(this.state.selectedOldDistrictId)]
                                : '#ccc',
                            display: 'block',
                            borderRadius: '50%'
                        }}
                    ></div>
                    <DropdownButton
                        id="dropdown-basic-button"
                        title={`Selected District: ${this.state.selectedOldDistrictId || 'N/A'}`}
                    >
                        {this.state.oldSortedKeys.length > 0 &&
                            this.state.oldSortedKeys.map(key => {
                                return (
                                    <Dropdown.Item key={key} onClick={() => this.selectOldDistrictData(key)}>
                                        District {key}
                                    </Dropdown.Item>
                                );
                            })}
                    </DropdownButton>
                </div>

                <br />

                {this.props.oldClusterCount && (
                    <Table size='sm' striped bordered hover>
                        <thead>
                            <tr>
                                <th>Election</th>
                                <th>Democratic District #</th>
                                <th>Republican District #</th>
                                <th>Tie District #</th>
                            </tr>
                        </thead>
                        <tbody>
                            <tr>
                                <td>{EnumNameMapper.getElectionName(this.props.oldClusterCount.election)}</td>
                                <td>{this.props.oldClusterCount.democraticCount}</td>
                                <td>{this.props.oldClusterCount.republicanCount}</td>
                                <td>{this.props.oldClusterCount.tieCount}</td>
                            </tr>
                        </tbody>
                    </Table>
                )}

                {this.state.oldDistrictData ? (
                    <div>
                        <p>
                            Total Population:{' '}
                            {this.state.oldDistrictData.demographicData.totalPopulation.toLocaleString()}
                        </p>
                        <StatisticsAccordionComponent
                            demographicData={this.state.oldDistrictData.demographicData}
                            electionData={this.state.oldDistrictData.electionData}
                            scores={this.state.oldDistrictData.objectiveFunctionScores}
                        />
                    </div>
                ) : (
                    <div className="text-center">
                        <h4>No district selected.</h4>
                        <h6>Select a district to view its data.</h6>
                    </div>
                )}

                <br />

                <h6>New District Statistics</h6>
                <div className="d-flex align-items-center">
                    <div
                        className="mr-2"
                        style={{
                            width: '28px',
                            height: '28px',
                            backgroundColor:
                                this.state.selectedNewDistrictId && this.state.selectedNewDistrictId !== '0'
                                    ? this.props.coloring.colors[parseInt(this.state.selectedNewDistrictId)]
                                    : '#ccc',
                            display: 'block',
                            borderRadius: '50%'
                        }}
                    ></div>
                    <DropdownButton
                        id="dropdown-basic-button"
                        title={`Selected District: ${
                            this.state.selectedNewDistrictId && this.state.selectedNewDistrictId !== '0'
                                ? this.state.selectedNewDistrictId
                                : 'N/A'
                        }`}
                    >
                        {this.state.newSortedKeys.length > 0 &&
                            this.state.newSortedKeys.map(key => {
                                return (
                                    <Dropdown.Item key={key} onClick={() => this.selectNewDistrictData(key)}>
                                        District {key} {this.isDistrictMajMin(key) ? `(MajMin)` : ``}
                                    </Dropdown.Item>
                                );
                            })}
                    </DropdownButton>
                </div>

                <br />

                {this.props.newClusterCount && (
                    <Table size='sm' striped bordered hover>
                        <thead>
                            <tr>
                                <th>Election</th>
                                <th>Democratic District #</th>
                                <th>Republican District #</th>
                                <th>Tie District #</th>
                            </tr>
                        </thead>
                        <tbody>
                            <tr>
                                <td>{EnumNameMapper.getElectionName(this.props.newClusterCount.election)}</td>
                                <td>{this.props.newClusterCount.democraticCount}</td>
                                <td>{this.props.newClusterCount.republicanCount}</td>
                                <td>{this.props.newClusterCount.tieCount}</td>
                            </tr>
                        </tbody>
                    </Table>
                )}

                {this.state.newDistrictData ? (
                    <div>
                        <p>
                            Total Population:{' '}
                            {this.state.newDistrictData.demographicData.totalPopulation.toLocaleString()}
                        </p>
                        {this.state.newDistrictData.isDistrictMajMin && <p>This is a majority/minority district.</p>}
                        <StatisticsAccordionComponent
                            demographicData={this.state.newDistrictData.demographicData}
                            electionData={this.state.newDistrictData.electionData}
                            scores={this.state.newDistrictData ? this.state.newDistrictData.objectiveFunctionScores : null}
                        />
                    </div>
                ) : (
                    <div className="text-center">
                        <h4>No district selected.</h4>
                        <h6>Select a district to view its data.</h6>
                    </div>
                )}

                <br />
            </div>
        );
    }

    private selectOldDistrictData(districtId: string) {
        this.setState({
            selectedOldDistrictId: districtId,
            oldDistrictData: this.props.oldClusters.get(districtId)
        });
    }

    private selectNewDistrictData(districtId: string) {
        this.setState({
            selectedNewDistrictId: districtId,
            newDistrictData: this.props.newClusters.get(districtId)
        });
    }

    private isDistrictMajMin(id: string): boolean {
        return this.props.newClusters.get(id).isMajorityMinority;
    }

    private resetDistrictResults() {
        this.setState({
            oldDistrictData: null,
            selectedOldDistrictId: '',
            oldSortedKeys: [],
            newDistrictData: null,
            selectedNewDistrictId: '',
            newSortedKeys: []
        });
    }
}
