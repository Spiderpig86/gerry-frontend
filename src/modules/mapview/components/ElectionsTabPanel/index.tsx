import * as React from 'react';

export interface IElectionsTabProps {
    presidentialResults: IVoteData;
    senatorialResults: IVoteData;
    gubernatorialResults: IVoteData;
}

export interface IVoteData {
    democraticVotes: number;
    republicanVotes: number;
    independentVotes: number;
}

export class ElectionsTabPanel extends React.PureComponent<{}, {}> {
    render() {
        return (
            <div style={{ padding: '0 1.5rem' }}>
                <br/>
                <h4>2016 Election</h4>
                <p><b>Presidential</b></p>
                <p>Democratic Votes:</p>
                <p>Republican Votes:</p>
                <p>Independent Votes:</p>
                <br/>
                <p><b>Senatorial</b></p>
                <p>Democratic Votes:</p>
                <p>Republican Votes:</p>
                <p>Independent Votes:</p>
                <br/>
                <p><b>Gubernatorial</b></p>
                <p>Democratic Votes:</p>
                <p>Republican Votes:</p>
                <p>Independent Votes:</p>
            </div>
        )
    }
}