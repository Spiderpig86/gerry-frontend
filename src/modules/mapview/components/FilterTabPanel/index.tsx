import * as React from 'react';
import * as Constants from '../../../../config/constants';
import * as mapActionCreators from '../../../../redux/modules/state/state';

import { Form } from 'react-bootstrap';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';

import { MapFilterEnum, ViewLevelEnum, FilterArgs } from '../../../../models';

interface IFilterProps {
    filterArgs: FilterArgs;
    setMapFilter: (filter: string) => void;
    setMapLevel: (level: string) => void;
}

class FilterTabPanelComponent extends React.Component<IFilterProps, {}> {

    render() {
        return (
            <div className='px-4 py-3'>
                <h4>Map Filter</h4>
                <h5>Display Level</h5>
                <Form className='mb-3'>
                    <Form.Group className="w-100 py-2 row form-group d-flex align-items-center">
                        <Form.Check defaultChecked={this.props.filterArgs.viewLevel === ViewLevelEnum.PRECINCTS} name='displayLevel' custom type={'radio'} id={`displayLevelPrecinct`} label={`Precincts`} onClick={() => this.props.setMapLevel(ViewLevelEnum.PRECINCTS)} />
                    </Form.Group>
                    <Form.Group className="w-100 py-2 row form-group d-flex align-items-center">
                        <Form.Check defaultChecked={this.props.filterArgs.viewLevel === ViewLevelEnum.OLD_DISTRICTS} name='displayLevel' custom type={'radio'} id={`displayLevelOldCd`} label={`Old Congressional Districts`} onClick={() => this.props.setMapLevel(ViewLevelEnum.OLD_DISTRICTS)} />
                    </Form.Group>
                    <Form.Group className="w-100 py-2 row form-group d-flex align-items-center">
                        <Form.Check defaultChecked={this.props.filterArgs.viewLevel === ViewLevelEnum.NEW_DISTRICTS} name='displayLevel' custom type={'radio'} id={`displayLevelNewCd`} label={`New Congressional Districts`} onClick={() => this.props.setMapLevel(ViewLevelEnum.NEW_DISTRICTS)} />
                    </Form.Group>
                </Form>
                
                <h6>Display Misc.</h6>
                <Form.Group className="w-100 py-2 row form-group d-flex align-items-center">
                    <Form.Check defaultChecked={this.props.filterArgs.mapFilter === MapFilterEnum.DEFAULT} name='population' custom type={'radio'} id={`displayDefault`} label={`Default Filter`} onClick={() => this.props.setMapFilter(MapFilterEnum.DEFAULT)} />
                </Form.Group>
                <br />

                <h6>Display Election Data</h6>
                <Form.Group className="w-100 py-2 row form-group d-flex align-items-center">
                    <Form.Check defaultChecked={this.props.filterArgs.mapFilter === MapFilterEnum.PRES_2016} name='population' custom type={'radio'} id={`electionPres16`} label={`2016 Presidential Election`} onClick={() => this.props.setMapFilter(MapFilterEnum.PRES_2016)} />
                </Form.Group>
                <Form.Group className="w-100 py-2 row form-group d-flex align-items-center">
                    <Form.Check defaultChecked={this.props.filterArgs.mapFilter === MapFilterEnum.HOUSE_2016} name='population' custom type={'radio'} id={`electionHouse16`} label={`2016 House Election`} onClick={() => this.props.setMapFilter(MapFilterEnum.HOUSE_2016)} />
                </Form.Group>
                {/* <Form.Group className="w-100 py-2 row form-group d-flex align-items-center">
                    <Form.Check defaultChecked={this.props.filterArgs.mapFilter === MapFilterEnum.SENATE_2016} name='population' custom type={'radio'} id={`electionSenate16`} label={`2016 Senate Election`} onClick={() => this.props.setMapFilter(MapFilterEnum.SENATE_2016)} />
                </Form.Group> */}
                <Form.Group className="w-100 py-2 row form-group d-flex align-items-center">
                    <Form.Check defaultChecked={this.props.filterArgs.mapFilter === MapFilterEnum.HOUSE_2018} name='population' custom type={'radio'} id={`electionHouse18`} label={`2018 House Election`} onClick={() => this.props.setMapFilter(MapFilterEnum.HOUSE_2018)} />
                </Form.Group>
                {/* <Form.Group className="w-100 py-2 row form-group d-flex align-items-center">
                    <Form.Check defaultChecked={this.props.filterArgs.mapFilter === MapFilterEnum.SENATE_2018} name='population' custom type={'radio'} id={`electionSenate18`} label={`2018 Senate Election`} onClick={() => this.props.setMapFilter(MapFilterEnum.SENATE_2018)} />
                </Form.Group> */}
                <br />

                <h6>Display Minority Population Density</h6>
                <Form.Group className="w-100 py-2 row form-group d-flex align-items-center">
                    <Form.Check defaultChecked={this.props.filterArgs.mapFilter === MapFilterEnum.WHITE_DENSITY} name='population' custom type={'radio'} id={`whiteDensity`} label={`Non-Hispanic White Density`} onClick={() => this.props.setMapFilter(MapFilterEnum.WHITE_DENSITY)} />
                </Form.Group>
                <Form.Group className="w-100 py-2 row form-group d-flex align-items-center">
                    <Form.Check defaultChecked={this.props.filterArgs.mapFilter === MapFilterEnum.HISPANIC_DENSITY} name='population' custom type={'radio'} id={`hispanicDensity`} label={`Hispanic Density`} onClick={() => this.props.setMapFilter(MapFilterEnum.HISPANIC_DENSITY)} />
                </Form.Group>
                <Form.Group className="w-100 py-2 row form-group d-flex align-items-center">
                    <Form.Check defaultChecked={this.props.filterArgs.mapFilter === MapFilterEnum.BLACK_DENSITY} name='population' custom type={'radio'} id={`blackDensity`} label={`Non-Hispanic African American Density`} onClick={() => this.props.setMapFilter(MapFilterEnum.BLACK_DENSITY)} />
                </Form.Group>
                <Form.Group className="w-100 py-2 row form-group d-flex align-items-center">
                    <Form.Check defaultChecked={this.props.filterArgs.mapFilter === MapFilterEnum.ASIAN_DENSITY} name='population' custom type={'radio'} id={`asianDensity`} label={`Non-Hispanic Asian Density`} onClick={() => this.props.setMapFilter(MapFilterEnum.ASIAN_DENSITY)} />
                </Form.Group>
                <Form.Group className="w-100 py-2 row form-group d-flex align-items-center">
                    <Form.Check defaultChecked={this.props.filterArgs.mapFilter === MapFilterEnum.NATIVE_AMERICAN_DENSITY} name='population' custom type={'radio'} id={`nativeAmericanDensity`} label={`Non-Hispanic Native American Density`} onClick={() => this.props.setMapFilter(MapFilterEnum.NATIVE_AMERICAN_DENSITY)} />
                </Form.Group>
                <Form.Group className="w-100 py-2 row form-group d-flex align-items-center">
                    <Form.Check defaultChecked={this.props.filterArgs.mapFilter === MapFilterEnum.PACIFIC_ISLANDER_DENSITY} name='population' custom type={'radio'} id={`pacificIslanderDensity`} label={`Non-Hispanic Pacific Islander Density`} onClick={() => this.props.setMapFilter(MapFilterEnum.PACIFIC_ISLANDER_DENSITY)} />
                </Form.Group>
                <Form.Group className="w-100 py-2 row form-group d-flex align-items-center">
                    <Form.Check defaultChecked={this.props.filterArgs.mapFilter === MapFilterEnum.OTHER_DENSITY} name='population' custom type={'radio'} id={`otherDensity`} label={`Other Density`} onClick={() => this.props.setMapFilter(MapFilterEnum.OTHER_DENSITY)} />
                </Form.Group>
                <Form.Group className="w-100 py-2 row form-group d-flex align-items-center">
                    <Form.Check defaultChecked={this.props.filterArgs.mapFilter === MapFilterEnum.BIRACIAL_DENSITY} name='population' custom type={'radio'} id={`biracialDensity`} label={`Biracial Density`} onClick={() => this.props.setMapFilter(MapFilterEnum.BIRACIAL_DENSITY)} />
                </Form.Group>
            </div>
        )
    }
}

function mapStateToProps(state: any) {
    return {
        filterArgs: state.stateReducer.filterArgs,
        setMapFilter: state.stateReducer.setMapFilter,
        setViewLevel: state.stateReducer.setViewLevel
    }
}

export const FilterTabPanel = connect(
    (state: any) => {
        return mapStateToProps(state);
    },
    dispatch => bindActionCreators(mapActionCreators, dispatch)
)(FilterTabPanelComponent);