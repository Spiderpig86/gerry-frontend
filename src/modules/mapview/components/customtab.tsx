import * as React from 'react';

import { Tab } from 'react-tabs';

export const CustomTab = ({ children }) => (
    <Tab>
      <h6>{children}</h6>
    </Tab>
);
// CustomTab.tabsRole = 'Tab'; // Required field to use your custom Tab