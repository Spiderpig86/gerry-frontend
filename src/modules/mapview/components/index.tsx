export { CustomTab } from './CustomTab';
export { LeftSidebar } from './LeftSidebar';
export { InputTabPanel } from './InputTabPanel';
export { LogsTabPanel } from './LogsTabPanel';
export { RightSidebar } from './RightSidebar';
export { AlgorithmPanel } from './AlgorithmPanel';