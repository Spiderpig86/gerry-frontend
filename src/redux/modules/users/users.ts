// Copyright (c) Microsoft. All rights reserved.
// Licensed under the MIT license. See LICENSE file in the project root for full license information.
import { User } from "models";
import { authenticate, checkIfAuthenticated } from "components";

const AUTH_USER = 'AUTH_USER'
const UNAUTH_USER = 'UNAUTH_USER'
const FETCHING_USER = 'FETCHING_USER'
const FETCHING_USER_FAILURE = 'FETCHING_USER_FAILURE'
const FETCHING_USER_SUCCESS = 'FETCHING_USER_SUCCESS'

const authUser = (uid: string) => {
  return {
    type: AUTH_USER,
    uid,
  }
}

export const unauthUser = () => {
  return {
    type: UNAUTH_USER,
  }
}

const fetchingUser = () => {
  return {
    type: FETCHING_USER,
  }
}

const fetchingUserFailure = (error: string) => {
  return {
    type: FETCHING_USER_FAILURE,
    error: 'Error fetching user.',
  }
}

const fetchingUserSuccess  = (uid: string, user: User, timestamp: Date) => {
  return {
    type: FETCHING_USER_SUCCESS,
    uid,
    user,
    timestamp,
  }
}

export const fetchAndHandleAuthentication = (history: any) => {
  return (dispatch: any) => {
    dispatch(fetchingUser());
    authenticate().then((user: any) => {
      dispatch(fetchingUserSuccess(user.uid, user, new Date()));
      dispatch(authUser(user.uid));
      history.push("/dashboard");
    }).catch((error) => dispatch(fetchingUserFailure(error)));
  };
}

const initialUserState: User = {
    uid: "",
    name: "",
    lastUpdated: new Date()
}

export const user = (state = initialUserState, action: any) => {
  switch (action.type) {
    case FETCHING_USER_SUCCESS :
      return {
        ...state,
        uid: action.uid,
        name: action.user.name,
        lastUpdated: action.timestamp
      }
    default :
      return state
  }
}

const initialState = {
  isFetching: false,
  error: "",
  isAuthed: false,
  authedId: "",
}

export const users = (state = initialState, action: any) => {
  switch (action.type) {
    case AUTH_USER :
      return {
        ...state,
        isAuthed: true,
        authedId: action.uid,
      }
    case UNAUTH_USER :
      return {
        ...state,
        isAuthed: false,
        authedId: "",
      }
    case FETCHING_USER:
      return {
        ...state,
        isFetching: true,
      }
    case FETCHING_USER_FAILURE:
      return {
        ...state,
        isFetching: false,
        error: action.error,
      }
    case FETCHING_USER_SUCCESS:
      return action.user === null
        ? {
          ...state,
          isFetching: false,
          error: "",
        }
        : {
          ...state,
          isFetching: false,
          error: "",
          [action.uid]: user(state[action.uid], action),
        }
    default :
      return state
  }
}