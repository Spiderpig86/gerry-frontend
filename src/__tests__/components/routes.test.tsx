// Copyright (c) Microsoft. All rights reserved.
// Licensed under the MIT license. See LICENSE file in the project root for full license information.
import React from "react";
import renderer from "react-test-renderer";
import { Routes } from "components";
import { createStore, applyMiddleware } from "redux";
import { Provider } from "react-redux";
import thunk from "redux-thunk";
import { users } from "../../redux/modules/users/users"

describe("Routes component tests", () => {
    it("Check Routes Component Render", () => {
        const store = createStore(users, applyMiddleware(thunk));
        const component = renderer.create(<Provider store={store}><Routes /></Provider>).toJSON();
        expect(component).toMatchSnapshot();
    });
});