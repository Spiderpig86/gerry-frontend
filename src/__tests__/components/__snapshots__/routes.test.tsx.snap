// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Routes component tests Check Routes Component Render 1`] = `
<div
  className="wrapper"
>
  <div
    className="navigation"
  >
    <div
      className="navigation-logo"
    >
      <nav
        className="navbar navbar-default"
      >
        <div
          className="container-fluid"
        >
          <div
            className="navbar-header"
          >
            <button
              aria-expanded="false"
              className="navbar-toggle collapsed"
              data-target="#nav-logo"
              data-toggle="collapse"
              type="button"
            >
              <span
                className="sr-only"
              >
                Toggle navigation
              </span>
              <span
                className="icon-bar"
              />
              <span
                className="icon-bar"
              />
              <span
                className="icon-bar"
              />
            </button>
            <a
              className="navbar-brand"
              href="/"
            >
              <span>
                Brand
              </span>
            </a>
          </div>
          <div
            className="collapse navbar-collapse"
            id="nav-logo"
          >
            <ul
              className="nav navbar-nav navbar-right"
            >
              <li
                className="hidden-md hidden-lg"
              >
                <a
                  href="/"
                  onClick={[Function]}
                >
                  Brand
                </a>
              </li>
              <li>
                <a
                  href="#"
                  onClick={[Function]}
                >
                  Sign in
                </a>
              </li>
            </ul>
          </div>
        </div>
      </nav>
    </div>
    <div
      className="navigation"
    >
      <nav
        className="navbar navbar-default"
      >
        <div
          className="container-fluid"
        >
          <div
            className="navbar-header"
          >
            <a
              className="navbar-brand hidden-xs hidden-sm"
              href="/"
              onClick={[Function]}
            >
              React-Redux Starer
            </a>
          </div>
          <div
            className="collapse navbar-collapse"
            id="nav"
          >
            
          </div>
        </div>
      </nav>
    </div>
  </div>
  <div
    className="container-fluid"
  >
    <h1>
      404 Page Not Found
    </h1>
  </div>
  <div
    className="push"
  />
  <footer>
    <div
      className="container-fluid footer"
    >
      <div
        className="row"
      >
        <div
          className="col-md-12"
        >
          <ul
            className="pull-right list-unstyled"
          >
            <li>
              <a
                href=""
              >
                Contact us
              </a>
            </li>
            <li>
              Â© Microsoft 
              2018
            </li>
          </ul>
        </div>
      </div>
    </div>
  </footer>
</div>
`;
